monai/data/image_reader.py

```diff
@@ -626,3 +630,143 @@ def _get_spatial_shape(self, img) -> np.ndarray:
+
+        Args:
+            data: file name or a list of file names to read.
+
+        """
+        img_: List = []
+
+        filenames: Sequence[str] = ensure_tuple(data)
+        for name in filenames:
+            img = self.wsi_reader(name)
+            if self.wsi_reader_name == "openslide":
+                img.shape = (img.dimensions[1], img.dimensions[0], 3)
+            img_.append(img)
+
+        return img_ if len(filenames) > 1 else img_[0]
+
+    def get_data(
+        self,
+        img_obj,
+        location: Tuple[int, int] = (0, 0),
+        size: Optional[Tuple[int, int]] = None,
+        level: int = 0,
+        dtype: DtypeLike = np.uint8,
+        grid_shape: Tuple[int, int] = (1, 1),
+        patch_size: Optional[int] = None,
+    ):
+        """
+        Extract regions as numpy array from WSI image and return them.
+
+        Args:
+            img:      a wsi_reader object loaded from a file, or list of CuImage objects
+            location: (x_min, y_min) tuple giving the top left pixel in the level 0 reference frame,
+                       or list of tuples (default=(0, 0))
+            size:     (height, width) tuple giving the region size, or list of tuples (default=(wsi_width, wsi_height))
+                        This is the size of image at the given level (`level`)
+            level:    the level number, or list of level numbers (default=0)
+
+        """
+        if size is None:
+            if location == (0, 0):
+                # the maximum size is set to WxH
+                size = (img_obj.shape[0] // (2 ** level), img_obj.shape[1] // (2 ** level))
+                print(f"Reading the whole image at level={level} with shape={size}")
+            else:
+                print("Size need to be provided!")
+                return
+        region = self._extract_region(img_obj, location=location, size=size, level=level, dtype=dtype)
+        if patch_size is None:
+            patches = region
+        else:
+            patches = self._extract_patches(
+                region, patch_size=(patch_size, patch_size), grid_shape=grid_shape, dtype=dtype
+            )
+        return patches
+
+    def _extract_region(
+        self,
+        img_obj,
+        size: Tuple[int, int],
+        location: Tuple[int, int] = (0, 0),
+        level: int = 0,
+        dtype: DtypeLike = np.uint8,
+    ):
+        # reverse the order of dimensions for size and location to be compatible with image shape
+        size = size[::-1]
+        location = location[::-1]
+        region = img_obj.read_region(location=location, size=size, level=level)
+        if self.wsi_reader_name == "openslide":
+            region = region.convert("RGB")
+        # convert to numpy
+        region = np.asarray(region, dtype=dtype)
+        # cuCalaraImage/OpenSlide: (H x W x C) -> torch image: (C X H X W)
+        region = region.transpose((2, 0, 1))
+        return region
```

Comment:
Could you please help also add the meta data? Refer to the PILReader: https://github.com/Project-MONAI/MONAI/blob/master/monai/data/image_reader.py#L598

Reply:
metadata from OpenSlide and cuCalraImage are not structured in the same way so they results in a very different output which makes the output of WSIReader depend on the library to use. I believe it would be better to have a separate PR for that once the development of cuClaraImage become stable.

Reply:
Hi @behxyz ,

I agree to add well-structured metadata later, but for this very first version, would be useful to add at least the `spatial_shape` and `original_channel_dim`. Some other transforms or handlers may need this metadata for additional logic.
You can refer to the PILReader: https://github.com/Project-MONAI/MONAI/blob/master/monai/data/image_reader.py#L599
And don't transpose axis here, use `EnsureChannelFirst` transform in the transform chain instead.
If we decide to make all the readers output `channel-first`, I will submit another PR to support it in the reader base class.
Others of this PR look good to me now.

Thanks very much.

Reply:
@Nic-Ma,
giving more thought into this, I found some confusion here. When we are talking about `spatial_shape` and `original_channel_dim`, is it about the output image (which is a subregion of WSI) or the whole slide image?

What is the use for this? Because if it is about the output image, in this case `original_channel_dim` is meaningless because since it depends on the how we are extracting it, and if it is out the original whole slide image, `spatial_shape` becomes useless.

Reply:
For now, I put the `spatial_shape` and `original_channel_dim` for extracted region, which makes the most sense.